greybox_generalize.py

```diff
@@ -0,0 +1,249 @@
+
+        # only generating upper triangular part
+        # loop over parameters
+        for i in range(self.n_parameters):
+            # loop over parameters from current parameter to end
+            for j in range(i, self.n_parameters):
+                # flatten (i,j)
+                ele_to_order[(i, j)] = count
+                # this tuple is the position of this input in the FIM
+                str_name = (i, j)
+
+                # if an initial FIM is given, we can initialize with these values
+                if self.initial_fim is not None:
+                    pyomo_block.inputs[str_name].value = self.initial_fim[str_name]
+
+                # if not given initial FIM, we initialize with an identity matrix
+                else:
+                    # identity matrix
+                    if i == j:
+                        pyomo_block.inputs[str_name].value = 1
+                    else:
+                        pyomo_block.inputs[str_name].value = 0
+
+                count += 1
+
+        self.ele_to_order = ele_to_order
+
+    def set_input_values(self, input_values):
+        """
+        Set the values of the inputs.
+
+        Arguments
+        ---------
+        input_values: input initial values
+        """
+        self._input_values = list(input_values)
+
+    def evaluate_equality_constraints(self):
+        """Evaluate the equality constraints.
+        Return None because there are no equality constraints.
+        """
+        return None
+
+    def evaluate_outputs(self):
+        """
+        Evaluate the output of the model.
+        We call numpy here to compute the logdet of FIM. slogdet is used to avoid ill-conditioning issue
+
+        Return
+        ------
+        logdet: a one-element numpy array, containing the log det value as float
+        """
+        # form matrix as a list of lists
+        M = self._extract_and_assemble_fim()
+
+        # compute log determinant
+        (sign, logdet) = np.linalg.slogdet(M)
+
+        if self.print_level >= 2:
+            print("iteration")
+            print("\n Consider M =\n", M)
+            print("Solution: ", self._input_values)
+            if self.print_level == 3:
+                print("   logdet = ", logdet, "\n")
+                print("Eigvals:", np.linalg.eigvals(M))
+
+        return np.asarray([logdet], dtype=np.float64)
+
+    def evaluate_jacobian_equality_constraints(self):
+        """Evaluate the Jacobian of the equality constraints."""
+        return None
+
+    def _extract_and_assemble_fim(self):
+        """
+        This function make the flattened inputs back into the shape of an FIM
+
+        Return
+        ------
+        M: a numpy array containing FIM.
+        """
+        # FIM shape Np*Np
+        M = np.zeros((self.n_parameters, self.n_parameters))
+        # loop over parameters.
+        # Alternately, ele_to_order only includes the upper triangle.
```

Comment:
Move the comment about ele_to_order containing only the upper triangular to inside the for loop

Reply:
moved